src/Microsoft.ML.StandardLearners/Standard/Online/OnlineLearnerStatic.cs

```diff
@@ -0,0 +1,169 @@
+using Microsoft.ML.StaticPipe.Runtime;
+using System;
+
+namespace Microsoft.ML.Trainers
+{
+    public static class OnlineLearnerStatic
+    {
+        /// <summary>
+        /// Predict a target using a linear binary classification model trained with the SDCA trainer, and a custom loss.
+        /// Note that because we cannot be sure that all loss functions will produce naturally calibrated outputs, setting
+        /// a custom loss function will not produce a calibrated probability column.
+        /// </summary>
+        /// <param name="ctx">The binary classification context trainer object.</param>
+        /// <param name="label">The label, or dependent variable.</param>
+        /// <param name="features">The features, or independent variables.</param>
+        /// <param name="lossFunction">The custom loss.</param>
+        /// <param name="weights">The optional example weights.</param>
+        /// <param name="learningRate">The learning Rate.</param>
+        /// <param name="decreaseLearningRate">Decrease learning rate as iterations progress.</param>
+        /// <param name="l2RegularizerWeight">L2 Regularization Weight.</param>
+        /// <param name="numIterations">Number of training iterations through the data.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}.Fit(DataView{TTupleInShape})"/> method is called on the
+        /// <see cref="Estimator{TTupleInShape, TTupleOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained, as well as the calibrator on top of that model. Note that this action cannot change the
+        /// result in any way; it is only a way for the caller to be informed about what was learnt.</param>
+        /// <returns>The set of output columns including in order the predicted binary classification score (which will range
+        /// from negative to positive infinity), and the predicted label.</returns>
+        /// <seealso cref="AveragedPerceptronTrainer"/>.
+        public static (Scalar<float> score, Scalar<bool> predictedLabel) AveragedPerceptron(
+                this BinaryClassificationContext.BinaryClassificationTrainers ctx,
+                IClassificationLoss lossFunction,
+                Scalar<bool> label, Vector<float> features, Scalar<float> weights = null,
+                float learningRate = AveragedLinearArguments.AveragedDefaultArgs.LearningRate,
+                bool decreaseLearningRate = AveragedLinearArguments.AveragedDefaultArgs.DecreaseLearningRate,
+                float l2RegularizerWeight = AveragedLinearArguments.AveragedDefaultArgs.L2RegularizerWeight,
+                int numIterations = AveragedLinearArguments.AveragedDefaultArgs.NumIterations,
+                Action<LinearBinaryPredictor> onFit = null
+            )
+        {
+            Contracts.CheckValue(label, nameof(label));
```

Comment:
>Contracts [](start = 12, length = 9)

technically you have ctx.Environment

Reply:
ctx is just the trainers, not the context. I don't have the env.

---
In reply to: [221115177](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221115177) [](ancestors = 221115177)

Reply:
Oh, you do. You need to call `StaticPipeUtils.GetEnvironment` to get it

---
In reply to: [221295921](https://github.com/dotnet/machinelearning/pull/1084#discussion_r221295921) [](ancestors = 221295921,221115177)